nk3: Add test subcommand

robin-nitrokey opened the pull request:
This patch adds the nk3 test subcommand that checks whether the
connected Nitrokey 3 devices work as expected.  It performs the
following steps:
- enumerate the CTAPHID devices
- query the firmware version
- query the UUID (if available)
- perform FIDO2 make credential and get assertion commands
----
Depends on #106.

robin-nitrokey commented:
Rebased onto master, should be ready for review now.  Please have a closer look at the FIDO2 operations and whether there are any more checks we should run.

szszszsz commented:
1. I think current implementation of the test does not handle devices with PIN set. Please check.
2. Please add x5c fingerprint verification as part of the test suite (from `fido2 verify` command), see below:
    https://github.com/Nitrokey/pynitrokey/blob/70bcdd8ecdae84cffa73f00924dc34d9d414f963/pynitrokey/cli/fido2.py#L579-L582
    https://github.com/Nitrokey/pynitrokey/blob/70bcdd8ecdae84cffa73f00924dc34d9d414f963/pynitrokey/cli/fido2.py#L627-L639

robin-nitrokey commented:
> I think current implementation of the test does not handle devices with PIN set. Please check.

That’s true, I considered that too complex for the first iteration and not necessary for the basic diagnosis.  Do you think this is necessary for the first implementation?

> Please add x5c fingerprint verification as part of the test suite (from fido2 verify command), see below:

Will do.  Can you provide me the expected fingerprint from the production tool?

szszszsz reviewed (changes requested):
Implementation is fine for a quick test tool, however ideally I wish this would be better structured, with each test being separate function, and executed from a test list rather than called directly (similarly to the pytest execution, not necessarily meaning automatic test discovery).

1. Please add changes as requested in the previous comment.
2. Please add FIXME comment marking to restructure this later.

szszszsz commented:
1. This depends at what stage user would like to diagnose his device. We can ask to factory reset it, but it might not work or be too hard for the user to execute. I am not sure we have implementation handling this though, so it is fine to release it as-is now, and improve in the next iteration.
2. Already added to pynitrokey. See https://github.com/Nitrokey/pynitrokey/pull/107#L629

szszszsz commented:
More ideas:
1. Check if the device is listed by the OS at all (e.g. using libusb / hidapi listings) or just calling `lsusb`.
2. Check for the udev file being in place - a couple of hardcoded paths to check containing the rule for Nitrokey 3 (extra: `pkgconfig udev --dir`, available rather only on the development machines though).

robin-nitrokey commented:
I’ve rebased the patch onto master, updated the type stubs and re-structured the tests as you suggested.  Can you please review again?  My suggestion would be to merge the current functionality and add more tests and log more information in separate PRs in the future.

robin-nitrokey commented:
For reference, here is the example output for two devices, one with a hardcoded failure (`raise ValueError("dummy error")`).  The log contains the full exception and traceback.

```
$ nitropy nk3 test
Nitrokey tool for Nitrokey FIDO2, Nitrokey Start, Nitrokey 3 & NetHSM
Found 2 Nitrokey 3 device(s):
- Nitrokey 3 at /dev/hidraw4
- Nitrokey 3 at /dev/hidraw3

Running tests for Nitrokey 3 at /dev/hidraw4

[1/3]   UUID query                      SUCCESS         DD59A7F00CCB715DB515F83C346CA8AD
[2/3]   Firmware version query          SUCCESS         v1.0.1
Please press the touch button on the device ...
Please press the touch button on the device ...
[3/3]   FIDO2                           SUCCESS         44fa598fdc98681dc5c8659a804c40bd6e53f8e54a781608b0651d47a53e1c8a

3 tests, 3 successful, 0 skipped, 0 failed

Running tests for Nitrokey 3 at /dev/hidraw3

[1/3]   UUID query                      FAILURE         dummy error
[2/3]   Firmware version query          SUCCESS         v1.0.0
Please press the touch button on the device ...
Please press the touch button on the device ...
[3/3]   FIDO2                           SUCCESS         None

3 tests, 2 successful, 0 skipped, 1 failed

Summary: 2 device(s) tested, 1 successful, 1 failed

Critical error:
Test failed for 1 device(s)

--------------------------------------------------------------------------------
Critical error occurred, exiting now
Unexpected? Is this a bug? Do you would like to get support/help?
- You can report issues at: https://github.com/Nitrokey/pynitrokey/issues/
- Writing an e-mail to: support@nitrokey.com is also possible
- Please attach the log: '/tmp/nitropy.log.cnb25e18' with any support/help request!
--------------------------------------------------------------------------------
```

szszszsz commented on pynitrokey/cli/nk3/test.py:
```diff
@@ -0,0 +1,189 @@
+
+        local_print(f"[{idx}/{n}]\t{name}\t{status}\t{msg}")
+
+    success = len([result for result in results if result.status == TestStatus.SUCCESS])
+    skipped = len([result for result in results if result.status == TestStatus.SKIPPED])
+    failed = len([result for result in results if result.status == TestStatus.FAILURE])
```
nit: you can use `filter()` here

robin-nitrokey replied:
thanks.  I was sure that there was a built-in function like that but I didn’t find it at first glance, so I wrote this. ;-)

robin-nitrokey replied:
Looks like `filter` does not help much here because it returns an iterable.  So we have to write:

```
    success = len(list(filter(lambda result: result.status == TestStatus.SUCCESS, results)))
```

which is not that much of an improvement.

szszszsz replied:
One can argue using `filter()` shows better the intent, and thus increases readability. I do not see in this case improvement either though. Leaving decision to you.

szszszsz commented on pynitrokey/cli/nk3/test.py:
```diff
@@ -0,0 +1,189 @@
+    local_print("")
+    local_print(f"{n} tests, {success} successful, {skipped} skipped, {failed} failed")
+
+    return not any([result.status == TestStatus.FAILURE for result in results])
```
nit: for readability and cognition load I think it would be better to check for `all()` instead of  `not any()`, "positive" logic so to speak

robin-nitrokey replied:
sure, I will update it

robin-nitrokey replied:
done

szszszsz commented on pynitrokey/cli/nk3/test.py:
```diff
@@ -0,0 +1,189 @@
+            result = TestResult(TestStatus.FAILURE, exc_info=sys.exc_info())
+        results.append(result)
+
+        idx = str(i + 1).rjust(idx_len)
+        name = test_case.name.ljust(name_len)
+        status = result.status.name.ljust(status_len)
```
nit: this could be in the format string

robin-nitrokey replied:
In this case I decided to separate it because format strings with multiple dynamic width specifiers are rather hard to read.

szszszsz commented on pynitrokey/cli/nk3/test.py:
```diff
@@ -0,0 +1,189 @@
+            )
+            msg = str(result.exc_info[1])
+
+        local_print(f"[{idx}/{n}]\t{name}\t{status}\t{msg}")
```
Note: this would be nice use case for tabulate, shall we have more views like this
- https://pypi.org/project/tabulate/

szszszsz commented on pynitrokey/cli/nk3/test.py:
```diff
@@ -0,0 +1,189 @@
+    if not isinstance(device, Nitrokey3Device):
+        return TestResult(TestStatus.SKIPPED)
+
+    # Based on https://github.com/Yubico/python-fido2/blob/142587b3e698ca0e253c78d75758fda635cac51a/examples/credential.py
```
Note: I think that if this is a verbatim copy, the whole copyrights remarks section have to be copied as well.

robin-nitrokey replied:
The similarities are just dictated by the API, so I don’t think this is necessary here, but I can add the Yubico copyright statement to the file header to be sure.

robin-nitrokey replied:
done

szszszsz reviewed (approved):
LGTM 
Please check license issues.

szszszsz commented:
Feel free to merge once updated
